api/types/role.go

```diff
@@ -612,6 +621,37 @@ func (r *RoleV3) CheckAndSetDefaults() error {
 	return nil
 }
 
+// DowngradeToV3 converts a V4 role to V3 so that it will be compatible with older instances.
+// DELETE IN 8.0.0
+func (r *RoleV3) DowngradeToV3() error {
+	switch r.Version {
+	case V3:
+	case V4:
+		r.Version = V3
+
+		// V3 roles will set the default labels to wildcard allow if they are
+		// empty. To prevent this for roles which are created as V4 and
+		// downgraded, set a placeholder label
+		const labelKey = "__teleport_no_labels"
+		labelVal := uuid.NewString()
+		if r.Spec.Allow.NodeLabels == nil || len(r.Spec.Allow.NodeLabels) == 0 {
```

Comment:
Here and below, checking `len(...) == 0` would be sufficient as it covers the "nil" case as well.

Reply:
good point, updated